### ISSUE TITLE
GUI becomes unresponsive after a few minutes

### ISSUE THREAD

**Issue by JamesK89** [NationalSecurityAgency/ghidra#1516]
**Describe the bug**
Ghidra initially runs fine in the first few minutes but eventually the GUI will completely lock up and will be unresponsive to keyboard or mouse input.

Attempting to close the application yields no response, so the process needs to be killed to close it.
The windows also do not redraw if you resize them while the application is in this state.

Normally when an application freezes up Windows will label the application as "Not Responding",  however that is not happening in this case.

Running Ghidra in debug mode via ghidraDebug.bat does not yield any unusual messages in the console.
Changing the MAXMEM environment  variable in ghidraRun.bat seems to have no impact on this particular problem.

**To Reproduce**
1. Start Ghidra normally.
2. When the project window opens, wait a few minutes.
3. Notice that the application is now unresponsive to any sort of user input.
4. Click the "X" in corner of the window and notice the application does not close.

**Expected behavior**
The GUI should be responsive even after running for a few minutes.

**Environment (please complete the following information):**
 - OS: Windows 10 Professional Kernel 10.0.17134.472
 - Java Version: JDK 13.0.2 / JRE 8.0.241
 - Ghidra Version: 9.1.1

**Additional context**
To narrow down this problem to see if it had to do with a specific project I had started Ghidra, closed my project, closed Ghidra and then started Ghidra again so there would be no project loaded.
The issue continued to occur even without a project loaded.

**Comment by dev747368** [comment]
Please try jdk 11.  (see installation guide for requirements)

**Comment by JamesK89** [comment]
I've completely purged my computer of all traces of Java and installed only JDK 11.0.6.
Despite this the problem continues to occur.

**Comment by dragonmacher** [comment]
A thread dump would be helpful.

Once you are in the bad state, from a terminal, run:

- `jps`  (to get the pid of the Ghidra process)
- `jstack` <pid> (to get the application thread state

**Comment by JamesK89** [comment]
Here you go:

[thread-dump.txt](https://github.com/NationalSecurityAgency/ghidra/files/4161525/thread-dump.txt)

**Comment by dragonmacher** [comment]
The Swing thread appears to be showing a Ghidra file chooser widget.   Does this only happen after showing a file chooser?

**Comment by JamesK89** [comment]
No, it happens even if I just start Ghidra and let it idle with no project loaded. The file chooser is very slow though to the point of being practically unusable outside of typing in the path of the file directly.

**Comment by dragonmacher** [comment]
I see.  Will you try the above again, this time without using a file chooser?

Are you running in a VM?  I suspect there is some sort of slow file system interaction between the JVM and the host OS.

**Comment by JamesK89** [comment]
Ok, here you go. This time I just let Ghidra idle with no open projects while occasionally opening a menu or resizing the window to see if it was still responsive. Once it became unresponsive I generated the attached thread dump.

[thread-dump-the-sequel.txt](https://github.com/NationalSecurityAgency/ghidra/files/4162279/thread-dump-the-sequel.txt)

**Comment by dragonmacher** [comment]
This is interesting.  The file chooser is still causing the issue.   

There is a plugin that loads file chooser instances as it initializes, even though it does not yet need them.   Our file chooser uses a background thread for loading directories.  In your thread dump we can see 2 background threads loading, as well as the Swing thread, which is why the UI is locked.

I can see from the traces that the code is bogging down here:
```
	at java.io.WinNTFileSystem.getBooleanAttributes(java.base@11.0.6/Native Method)
	at java.io.File.isDirectory(java.base@11.0.6/File.java:852)
	at docking.widgets.filechooser.LocalFileChooserModel.isDirectory(LocalFileChooserModel.java:238)
```
The code shown above is calling into the OS-native code to determine if a particular file is a directory.   This is happening during the sorting process and is thus called a great deal of times for each directory whose contents are being sorted within a Ghidra file chooser.

We have seen this before in the past when running a Windows VM with a non-Windows host.  We plan on making some changes to make the code a little less wasteful.  However, this will not fix your issue.

**Comment by dev747368** [comment]
In addition to what @dragonmacher  said, I've got a snippet of code you can run in a java shell on your machine that will help narrow down the issue.  This code basically just does a list of files in your home directory and sorts them in the same way the ghidra file chooser does, and then displays the time it took to do that.

At a cmd prompt, run jshell (you should see some verbage about "Welcome to JShell -- Version 11.0.3", and then a jshell> prompt), and then on one line, paste this long string:

`long start = System.currentTimeMillis(); File[] filelist = new File(System.getProperty("user.home")).listFiles(); Arrays.sort(filelist, (f1, f2) -> {int x = Boolean.compare(f1.isDirectory(), f2.isDirectory()); return x == 0 ? f1.getName().compareTo(f2.getName()) : x;} ); long elapsed = System.currentTimeMillis() - start; System.out.println("Files: " + filelist.length + ", elapsed: " + elapsed);`

Look for the output that says something like: `Files: 122, elapsed: 684` and share that with us here.

**Comment by JamesK89** [comment]
I believe I found the issue. I had an old symlink to a network share that I don't have on the network I am currently connected to. Windows Explorer even hung up when I tried to access it. As soon as I removed the symlink from the command line both Explorer, the test script and Ghidra became responsive.

With the case of the test script I was getting an elapsed time of 558996 but after removing the symlink the elapsed time was 143.

This issue might been happening in virtual machines because some virtual machines will have a network share in the home directory to allow file sharing between the host and the guest.
